rubygems/index.js

```diff
@@ -0,0 +1,230 @@
+      name = resolved.name
+      version = resolved.version
+    } catch (e) {
+      this.log.error(e)
+      // unable to resolve the given spec; no way to get the deps for this input
+      return []
+    }
+
+    let body
+    if (!version) {
+      ({ body } = await this.got(`https://rubygems.org/api/v2/rubygems/${name}.json`, {
+        responseType: "json"
+      }))
+    } else {
+      ({ body } = await this.got(
+        `https://rubygems.org/api/v2/rubygems/${name}/versions/${version}.json`,
+        {
+          responseType: "json"
+        }
+      ))
+    }
+
+    const dependencyKeys = Object.keys(body.dependencies)
+    const depRequirements = dependencyKeys.reduce((acc, key) => {
+      const runtime = body.dependencies[key]
+      return acc.concat(runtime.reduce((ac, dep) => {
+        // each dep is formatted like this
+        /**
+         * {
+              "name": "activerecord",
+              "requirements": "= 3.0.18"
+          }
+        */
+        const versionSplit = dep.requirements.split(' ')
+        const spec = {
+          name: dep.name,
+          version: versionSplit[1],
+          operator: versionSplit[0],
+          toString: () => `${dep.name}@${versionSplit[0]}${versionSplit[1]}`
+        }
+        return ac.concat(spec)
+      }, []))
+    }, [])
+
+    if (!depRequirements || !depRequirements.length) return []
+
+    return depRequirements
+  }
+
+  // input: output of getSpec
+  // resolves the most suitable version to freeze given <name><operator><version>
+  // for example: rubocop >= 3.0.0 would fetch all versions available, and select the highest version
+  async resolve(pkgSpec) {
+    const { name, operator, versionSpec: version } = pkgSpec
+
+    if (operator === "==" || operator === '=') {
+      return { name, version }
+    }
+
+    // Fetch all tags for a package from https://rubygems.org/api/v1/versions/[gem name].json . 
+    // response will be an array of releases with a "number" key
+    const { body } = (await this.got(`https://rubygems.org/api/v1/versions/${name}.json`,
+      {
+        responseType: "json"
+      }
+    ))
+    // Grab releases and sort them greatest to least
+    const releases = body.map((rel) => rel.number)
+      .sort(compareVersions)
+      .reverse()
+
+    if (!releases.length) throw new Error("No releases found")
+
+    // If version is latest, return the first element of the releases array, representing the latest release
+    if (version === "latest") {
+      return { name, version: releases.shift() }
+    }
+
+    let release
+
+    // Find the highest version out of the tags that satisfy the requirements
+    switch (operator) {
+      case ">=":
+      case "<=":
+      case ">":
+        release = releases.find(rel =>
+          compareVersions.compare(rel, version, operator)
+        )
+        break
+      case "~>":
+        // Fetch latest version under the next specified minor
+        const versionComponents = version.split('.')
```

Comment:
wondering how we can make this block more easily understood..........no ideas currently

Reply:
ya... :(